# charon integration step 40 after 546078ea5d: commit ac69a1513c8707bea69389b4794e41ef1032e5b3
# first parent 9b4eec39818095d046fa64a02af3420147e16cf3
# series charon: patch -p1 in charon/ of the upstream source
#
# charon: --hide-marker-traits hides the auto traits of the standard library
#
# Send, Sync, Unpin, Freeze, UnwindSafe and RefUnwindSafe join the hidden marker traits: their
# clauses, built-in impls (BuiltinOrAuto with Auto data, which Aeneas does not handle), supertrait
# clauses and explicit impls disappear. Auto traits have no items, so no information the
# translated code uses is lost. UI test hide-auto-traits.

diff --git a/src/options.rs b/src/options.rs
index 649b464..d4da273 100644
--- a/src/options.rs
+++ b/src/options.rs
@@ -537,8 +537,8 @@ pub struct TranslateOptions {
     pub monomorphize_mut: Option<MonomorphizeMut>,
     /// Remove trait clauses attached to type declarations.
     pub remove_adt_clauses: bool,
-    /// Whether to hide various marker traits such as `*Sized` and `Destruct` anywhere they show
-    /// up.
+    /// Whether to hide various marker traits such as `*Sized`, `Destruct` and the auto traits
+    /// anywhere they show up.
     pub hide_marker_traits: bool,
     /// Hide the `A` type parameter on standard library containers (`Box`, `Vec`, etc).
     pub hide_allocator: bool,
@@ -654,6 +654,14 @@ impl TranslateOptions {
                 "core::marker::PointeeSized",
                 "core::marker::Tuple",
                 "core::clone::TrivialClone",
+                // The auto traits of the standard library. They have no items, so their clauses
+                // carry no information for the translated code.
+                "core::marker::Send",
+                "core::marker::Sync",
+                "core::marker::Unpin",
+                "core::marker::Freeze",
+                "core::panic::unwind_safe::UnwindSafe",
+                "core::panic::unwind_safe::RefUnwindSafe",
             ])
             .into_iter()
             .flatten()
diff --git a/tests/ui/hide-auto-traits.out b/tests/ui/hide-auto-traits.out
new file mode 100644
index 0000000..7fb116b
--- /dev/null
+++ b/tests/ui/hide-auto-traits.out
@@ -0,0 +1,253 @@
+# Final LLBC before serialization:
+
+opaque type core::any::Any::{vtable}
+
+// Full name: core::any::TypeId
+#[lang_item("type_id")]
+pub opaque type TypeId
+
+// Full name: core::any::Any
+#[lang_item("Any")]
+pub trait Any<Self>
+where
+    Self : 'static,
+{
+    fn type_id<'_0_1> = core::any::Any::type_id<'_0_1, Self>[Self]
+    vtable: core::any::Any::{vtable}
+}
+
+pub fn core::any::Any::type_id<'_0, Self>(_1: &'_0 Self) -> TypeId
+where
+    [@TraitClause0]: Any<Self>,
+= <opaque>
+
+// Full name: core::any::{impl Any for T}::type_id
+pub fn {impl Any for T}::type_id<'_0, T>(_1: &'_0 T) -> TypeId
+where
+    T : 'static,
+= <opaque>
+
+// Full name: core::any::{impl Any for T}
+impl<T> Any for T
+where
+    T : 'static,
+{
+    fn type_id<'_0_1> = {impl Any for T}::type_id<'_0_1, T>
+    vtable: {impl Any for T}::{vtable}<T>
+}
+
+// Full name: core::ops::function::FnOnce
+#[lang_item("fn_once")]
+pub trait FnOnce<Self, Args>
+{
+    type Output
+    fn call_once = core::ops::function::FnOnce::call_once<Self, Args>[Self]
+    vtable: core::ops::function::FnOnce::{vtable}<Args, Self::Output>
+}
+
+pub fn core::ops::function::FnOnce::call_once<Self, Args>(_1: Self, _2: Args) -> @TraitClause0::Output
+where
+    [@TraitClause0]: FnOnce<Self, Args>,
+= <opaque>
+
+// Full name: alloc::alloc::Global
+#[lang_item("global_alloc_ty")]
+pub struct Global {}
+
+// Full name: UNIT_METADATA
+fn UNIT_METADATA()
+{
+    let _0: (); // return
+
+    _0 = ()
+    return
+}
+
+// Full name: UNIT_METADATA
+const UNIT_METADATA: () = UNIT_METADATA()
+
+// Full name: test_crate::boxed
+pub fn boxed<P>(payload_1: P) -> alloc::boxed::Box<(dyn Any + Send + 'static)>
+where
+    [@TraitClause0]: Any<P>,
+{
+    let _0: alloc::boxed::Box<(dyn Any + Send + '2)>; // return
+    let payload_1: P; // arg #1
+    let _2: alloc::boxed::Box<(dyn Any + Send + '3)>; // anonymous local
+    let _3: alloc::boxed::Box<P>; // anonymous local
+    let _4: P; // anonymous local
+
+    storage_live(_2)
+    storage_live(_3)
+    storage_live(_4)
+    _4 = move payload_1
+    _3 = @BoxNew<P>(move _4)
+    _2 = unsize_cast<alloc::boxed::Box<P>, alloc::boxed::Box<(dyn Any + Send + '4)>, &vtable_of(@TraitClause0)>(move _3)
+    conditional_drop[{impl#2}<P, Global>] _3
+    storage_dead(_4)
+    storage_dead(_3)
+    _0 = unsize_cast<alloc::boxed::Box<(dyn Any + Send + '3)>, alloc::boxed::Box<(dyn Any + Send + '5)>,  at []>(move _2)
+    conditional_drop[{impl#2}<(dyn Any + Send + '9), Global>] _2
+    storage_dead(_2)
+    conditional_drop[{built_in impl Destruct for P}] payload_1
+    return
+}
+
+// Full name: test_crate::call
+pub fn call<'_0, F, T>(f_1: F, _t_2: &'_0 T) -> u32
+where
+    [@TraitClause0]: FnOnce<F, ()>,
+    @TraitClause0::Output = u32,
+{
+    let _0: u32; // return
+    let f_1: F; // arg #1
+    let _t_2: &'1 T; // arg #2
+    let _3: F; // anonymous local
+    let _4: (); // anonymous local
+
+    storage_live(_3)
+    _3 = move f_1
+    storage_live(_4)
+    _4 = ()
+    _0 = @TraitClause0::call_once(move _3, move _4)
+    storage_dead(_4)
+    storage_dead(_3)
+    conditional_drop[{built_in impl Destruct for F}] f_1
+    return
+}
+
+// Full name: test_crate::Job
+pub trait Job<Self>
+{
+    fn run<'_0_1> = test_crate::Job::run<'_0_1, Self>[Self]
+    vtable: test_crate::Job::{vtable}
+}
+
+pub fn test_crate::Job::run<'_0, Self>(_1: &'_0 Self) -> u32
+where
+    [@TraitClause0]: Job<Self>,
+= <method_without_default_body>
+
+// Full name: test_crate::run_job
+pub fn run_job<'_0, J>(job_1: &'_0 J) -> u32
+where
+    [@TraitClause0]: Job<J>,
+{
+    let _0: u32; // return
+    let job_1: &'1 J; // arg #1
+    let _2: &'2 J; // anonymous local
+
+    storage_live(_2)
+    _2 = &(*job_1)
+    _0 = @TraitClause0::run<'4>(move _2)
+    storage_dead(_2)
+    return
+}
+
+// Full name: test_crate::Handle
+pub struct Handle<T> {
+  T,
+}
+
+// Full name: test_crate::{impl Job for Handle<T>}::run
+pub fn {impl Job for Handle<T>}::run<'_0, T>(self_1: &'_0 Handle<T>) -> u32
+{
+    let _0: u32; // return
+    let self_1: &'1 Handle<T>; // arg #1
+
+    _0 = const 0 : u32
+    return
+}
+
+// Full name: test_crate::{impl Job for Handle<T>}
+impl<T> Job for Handle<T> {
+    fn run<'_0_1> = {impl Job for Handle<T>}::run<'_0_1, T>
+    vtable: {impl Job for Handle<T>}::{vtable}<T>
+}
+
+// Full name: test_crate::user::closure
+struct closure {}
+
+// Full name: test_crate::user::{impl FnOnce<()> for closure}::call_once
+fn {impl FnOnce<()> for closure}::call_once(_1: closure, tupled_args_2: ()) -> u32
+{
+    let _0: u32; // return
+    let _1: closure; // arg #1
+    let tupled_args_2: (); // arg #2
+
+    _0 = const 3 : u32
+    return
+}
+
+// Full name: test_crate::user::{impl FnOnce<()> for closure}
+impl FnOnce<()> for closure {
+    type Output = u32
+    fn call_once = {impl FnOnce<()> for closure}::call_once
+    non-dyn-compatible
+}
+
+// Full name: test_crate::user
+pub fn user() -> u32
+{
+    let _0: u32; // return
+    let _1: alloc::boxed::Box<(dyn Any + Send + '2)>; // anonymous local
+    let _2: u32; // anonymous local
+    let _3: closure; // anonymous local
+    let _4: &'4 u8; // anonymous local
+    let _5: &'5 u8; // anonymous local
+    let _6: u32; // anonymous local
+    let _7: &'7 Handle<u8>; // anonymous local
+    let _8: &'8 Handle<u8>; // anonymous local
+    let _9: u32; // anonymous local
+    let _10: &'9 Handle<u8>; // anonymous local
+    let _11: &'10 u8; // anonymous local
+    let _12: &'21 u8; // anonymous local
+    let _13: u8; // anonymous local
+    let _14: &'22 Handle<u8>; // anonymous local
+    let _15: Handle<u8>; // anonymous local
+
+    storage_live(_9)
+    storage_live(_10)
+    storage_live(_11)
+    storage_live(_1)
+    _1 = boxed<u32>[{impl Any for T}<u32>](const 1 : u32)
+    conditional_drop[{impl#2}<(dyn Any + Send + '14), Global>] _1
+    storage_live(_12)
+    storage_live(_13)
+    _13 = const 0 : u8
+    _12 = &_13
+    storage_dead(_1)
+    storage_live(_2)
+    storage_live(_3)
+    _3 = closure {  }
+    storage_live(_4)
+    storage_live(_5)
+    _11 = move _12
+    _5 = &(*_11)
+    _4 = &(*_5)
+    _2 = call<'17, closure, u8>[{impl FnOnce<()> for closure}](move _3, move _4)
+    storage_live(_14)
+    storage_live(_15)
+    _15 = Handle { 0: const 2 : u8 }
+    _14 = &_15
+    storage_dead(_4)
+    storage_dead(_3)
+    storage_live(_6)
+    storage_live(_7)
+    storage_live(_8)
+    _10 = move _14
+    _8 = &(*_10)
+    _7 = &(*_8)
+    _6 = run_job<'20, Handle<u8>>[{impl Job for Handle<T>}<u8>](move _7)
+    storage_dead(_7)
+    _9 = copy _2 panic.+ copy _6
+    _0 = move _9
+    storage_dead(_6)
+    storage_dead(_2)
+    storage_dead(_8)
+    storage_dead(_5)
+    return
+}
+
+
+
diff --git a/tests/ui/hide-auto-traits.rs b/tests/ui/hide-auto-traits.rs
new file mode 100644
index 0000000..6b84273
--- /dev/null
+++ b/tests/ui/hide-auto-traits.rs
@@ -0,0 +1,37 @@
+//@ charon-args=--hide-marker-traits
+//! Auto traits have no items: `--hide-marker-traits` hides them like the other marker traits, so
+//! that no clause, built-in impl, supertrait or explicit impl of `Send`, `Sync`, `Unpin`,
+//! `UnwindSafe` or `RefUnwindSafe` remains.
+use std::any::Any;
+use std::panic::{RefUnwindSafe, UnwindSafe};
+
+pub fn boxed<P: Any + Send>(payload: P) -> Box<dyn Any + Send> {
+    Box::new(payload)
+}
+
+pub fn call<F: FnOnce() -> u32 + UnwindSafe, T: Sync + Unpin + RefUnwindSafe>(f: F, _t: &T) -> u32 {
+    f()
+}
+
+pub trait Job: Send + Sync {
+    fn run(&self) -> u32;
+}
+
+pub fn run_job<J: Job>(job: &J) -> u32 {
+    job.run()
+}
+
+pub struct Handle<T>(pub T);
+
+unsafe impl<T: Send> Sync for Handle<T> {}
+
+impl<T: Send> Job for Handle<T> {
+    fn run(&self) -> u32 {
+        0
+    }
+}
+
+pub fn user() -> u32 {
+    let _ = boxed(1u32);
+    call(|| 3, &0u8) + run_job(&Handle(2u8))
+}
